cmd/cloud_sql_proxy/proxy.go

```diff
@@ -150,31 +137,177 @@ func listenInstance(dst chan<- proxy.Conn, dir, instance string) (net.Listener,
 				l.Close()
 				return
 			}
-			log.Printf("Got a connection for %q", instance)
-			dst <- proxy.Conn{instance, c}
+			log.Printf("Got a connection for %q", cfg.Instance)
+			dst <- proxy.Conn{cfg.Instance, c}
 		}
 	}()
 
-	log.Printf("Open socket for %q at %q", instance, path)
+	log.Printf("Open socket for %q at %q", cfg.Instance, cfg.Address)
 	return l, nil
 }
 
-// Check verifies that the dir parameter is set and that either 'fuse' is true or
-// at least one of instances/instancesSrc is set, but not both.
-func Check(dir string, fuse bool, instances []string, instancesSrc string) error {
-	switch {
-	case dir == "":
-		return errors.New("must set -dir")
-	case !fuse:
-		if len(instances) == 0 && instancesSrc == "" {
-			return errors.New("must specify -fuse, -instances, or -instances_metadata")
+type instanceConfig struct {
+	Instance         string
+	Network, Address string
+}
+
+// loopbackForNet maps a network (e.g. tcp6) to the loopback address for that
+// network. It is updated during the initialization of validNets to include a
+// valid loopback address for "tcp".
+var loopbackForNet = map[string]string{
+	"tcp4": "127.0.0.1",
+	"tcp6": "[::1]",
+}
+
+// validNets tracks the networks that are valid for this platform and machine.
+var validNets = func() map[string]bool {
+	m := map[string]bool{
+		"unix": runtime.GOOS != "windows",
+	}
+
+	anyTCP := false
+	for _, n := range []string{"tcp4", "tcp6"} {
+		addr, ok := loopbackForNet[n]
+		if !ok {
+			// This is effectively a compile-time error.
+			panic(fmt.Sprintf("no loopback address found for %v", n))
+		}
+		// Open any port to see if the net is valid.
+		x, err := net.Listen(n, addr+":")
+		if err != nil {
+			log.Printf("Protocol %v not supported: %v", n, err)
+			continue
+		}
+		x.Close()
+		m[n] = true
+
+		if !anyTCP {
+			anyTCP = true
+			// Set the loopback value for generic tcp if it hasn't already been
+			// set. (If both tcp4/tcp6 are supported the first one in the list
+			// (tcp4's 127.0.0.1) is used.
+			loopbackForNet["tcp"] = addr
+		}
+	}
+	if anyTCP {
+		m["tcp"] = true
+	}
+	return m
+}()
+
+func parseInstanceConfig(dir, instance string) (instanceConfig, error) {
+	var ret instanceConfig
+	eq := strings.Index(instance, "=")
+	if eq != -1 {
+		spl := strings.SplitN(instance[eq+1:], ":", 3)
+		ret.Instance = instance[:eq]
+
+		switch len(spl) {
+		default:
+			return ret, fmt.Errorf("invalid %q: expected 'project:instance=tcp:port'", instance)
+		case 2:
+			// No "host" part of the address. Be safe and assume that they want a
+			// loopback address.
+			ret.Network = spl[0]
+			addr, ok := loopbackForNet[spl[0]]
+			if !ok {
+				return ret, fmt.Errorf("invalid %q: unrecognized network %v", instance, spl[0])
+			}
+			ret.Address = fmt.Sprintf("%s:%s", addr, spl[1])
+		case 3:
+			// User provided a host and port; use that.
+			ret.Network = spl[0]
+			ret.Address = fmt.Sprintf("%s:%s", spl[1], spl[2])
```

Comment:
Got it. It looks like this makes #2 redundant with this change, right?  For maximum compatibility, would you want to do protocol:port(optional :address)? Or do you think it is more intuitive to do protocol:address:port? Thanks for adding in the option to specify what addresses to listen! Jason